### ISSUE TITLE
SymbolBlock error...??

### ISSUE THREAD

**Issue by nikkisingh111333** [deepjavalibrary/djl#956]
hey i m getting an error while building model for classification...when i first implement this i was not getting this error But nows i m getting this....
`java.lang.ClassCastException: class ai.djl.nn.SequentialBlock cannot be cast to class ai.djl.nn.SymbolBlock (ai.djl.nn.SequentialBlock and ai.djl.nn.SymbolBlock are in unnamed module of loader jdk.jshell.execution.DefaultLoaderDelegate$RemoteClassLoader `

heres my code very similar to what described in djl docs:

**ImageClassificationTranslator translator =
                ImageClassificationTranslator.builder()
                        .setPipeline(new Pipeline()
            .add(new Resize(416,416))
            .add(a -> a.div(255f))
            .add(new ToTensor())
                .optSynset(synset)
                        .optApplySoftmax(true)
                        .build();
Criteria<Image, Classifications> criteria = Criteria.builder()
    .setTypes(Image.class, Classifications.class)
    .optProgress(new ProgressBar())
    .optArtifactId("resnet")
    .optFilter("layers", "50")
    .optFilter("flavor", "v1")
    .optFilter("dataset", "cifar10").build();    
Model model = ModelZoo.loadModel(criteria);
System.out.println("Lond");
SequentialBlock newBlock = new SequentialBlock();
System.out.println(model.describeOutput().keys()+"----"+model.describeOutput().values());
SymbolBlock block =  (SymbolBlock)model.getBlock();
block.setInitializer( new NormalInitializer());
block.removeLastBlock();
newBlock.add(block);
newBlock.add(Blocks.batchFlattenBlock());
newBlock.add(Linear.builder().setUnits(256).build());
newBlock.add(Activation.reluBlock());
newBlock.add(Linear.builder().setUnits(128).build());
newBlock.add(Activation.reluBlock());
newBlock.add(Linear.builder().setUnits(2).build());
model.setBlock(newBlock);
model**

I already tried this to train cifar 10 where this was not an issue... i dont know why its giving me now...pleaase do help me?

**Comment by zachgk** [comment]
It might be that the Criteria is loading a different model. Try running `ModelZoo.listModels(criteria)` for your criteria. If this contains multiple models, it might mean it is giving you a different model than the one you are expecting. The loadModel has the same search, but stops after it finds the first model.

If you have added any DJL Model-Zoo dependencies, it could result in having more models available to be loaded and a different one actually loaded.

**Comment by nikkisingh111333** [comment]
Okk now training has done model loaded !!.but...whatever image i put it always give me one result.. i mean ..it always return first class as output....why is that..? i m guessing i m doing transfer learning on imagenet Weights which are already trained..how can i just train this predefined network on my dataset without using imagenet trained weights..i m assuming SymbolBlock.removeLastBlock() only remove last FC layers And Output but other than that major Cnn part is pre trained on imagenet...how can i just train resnet on my dataset intead of taking levearage imagenet weights...please explain me a bit??

**Comment by zachgk** [comment]
When you train, you improve the weights so the model performs better on your dataset. This requires that there are starting weights.

Usually, starting weights are created randomly using an `Initializer`. Another options is to use starting weights from training on another dataset such as Imagenet. If the dataset has any similarities to your dataset, it can make it train faster and more reliably because the starting weights are closer to what you need. So, I recommend using the imagenet weights since you already have them. See [this section of the D2L book](http://d2l.ai/chapter_computer-vision/fine-tuning.html?highlight=transfer%20learning) form more details. If you don't want this, just use a custom block and run `trainer.initialize`.

If your model is always returning the first class, are the probabilities changing or are they constant?

**Comment by nikkisingh111333** [comment]
yes probability of first class is always 0.99....no matter what image i use to predict........i m a little confuse about something in transfer learning..
please clarify me...
.My dataset has hand gestures images like one finger for label 1 and two finger for 2 each are just 50x50x3 dimentions..So if I train my network on Imagenet which i m doing using resnet18 with imagenet..and after using Symbole.removelastBlock()
So my real question is..
1.ImageNet has hand images trained weight ..so after training on ..will the hand weights help in my gesture dataset recognization..
2.if i run train.initialize() will evrything on resnet18 or whatever backbone architecture i m using train from scratch and get result what i want...How many epochs and Hyperparam do i needed...please help me ?
Currently i m also building vgg16 from scratch which is too time consuming..but still i m doing it..may i dont know Good use of transfer learning ..???

**Comment by zachgk** [comment]
Try passing in something like `manager.zeros(...)` and `manager.ones(...)`. Does that change it?

When training on something like imagenet, it learns how to detect items like cars or boats near the end layers which don't apply to your model. But, the early layers detect things like lines, shapes, and colors that are applicable. It saves your model some time as it doesn't have to relearn to detect these basic primitives.

You need to figure out the epochs and hyperparameters yourself. Try to experiment with different values and see what works. For epochs, train until the learning plateaus. Other hyperparameters you will just need to test.

We have an implementation of [VGG](https://github.com/deepjavalibrary/djl/blob/master/model-zoo/src/main/java/ai/djl/basicmodelzoo/cv/classification/VGG.java) already made. You can use it, but Resnet is a better model than VGG.

**Comment by nikkisingh111333** [comment]
so is there any way i can just take vgg16 model ..and train on my dataset instead using pre trained weights..i dont want weights of imagenet ....
can i extract layers and trained them on my dataset for example..is there any way i can train my network from 4th or 5th conv2d layer ..rest of the layers before 5th will be untrained and layers occuring after 5th will get triained on my dataset...if it could..i can i do it ??...

**Comment by zachgk** [comment]
You don't keep the pre-trained weights. As an analogy, imagine if you are traveling to New York. Using random weights is like starting really far away in San Fransisco. Using imagenet weights is like starting in Chicago. You still change all the weights to the New York weights (or the weights for your dataset), you just have less distance to travel to get there.

If you don't want that, use the VGG class I already mentioned. It doesn't load any weights from somewhere else and you don't need to do anything special with it.

**Comment by nikkisingh111333** [comment]
I have already implement VGG class...But it take a lot to create a model from scratch.....I just wanted model which we obtain using model zoo ..And on criteria instead of using using dataset="imagenet/cifar10"..Can choose to pass value null or none ..Means no pre trained weights to start ... just plain model...Then i ll just start train  on my dataset..Using image folder.is this possible?
